### ISSUE TITLE
Quarkus 2.x has java.lang.NoClassDefFoundError: Could not initialize class javax.crypto.JceSecurity

### ISSUE THREAD

**Issue by dhoffer** [quarkusio/quarkus#19265]
### Describe the bug

After upgrading from 1.13.7.Final to any of the 2.x versions we get this error at startup.

I retested with 2.1.1.Final and it has the same behavior.

We are using JDK 11.0.11+9.  Code and JDK work fine with pre 2.x Quarkus.

This code is just trying to decrypt a password in the io.quarkus.credentials.CredentialsProvider implementation used to connect to the database.

2021-08-05 15:21:21,296 ERROR [io.qua.run.Application] (main) Failed to start application (with profile prod): java.lang.NoClassDefFoundError: Could not initialize class javax.crypto.JceSecurity
        at java.base/javax.crypto.Cipher.getInstance(Cipher.java:540)
        at com.bs.util.encryption.AESEncryptDecrypt.getInitializedCipher(AESEncryptDecrypt.java:151)
        at com.bs.util.encryption.AESEncryptDecrypt.decrypt(AESEncryptDecrypt.java:127)
        at com.bs.util.encryption.AESEncryptDecryptProxy.decrypt(AESEncryptDecryptProxy.java:44)
        at com.bs.util.BSDataSourceCredentialsProvider.getCredentials(BSDataSourceCredentialsProvider.java:33)
        at com.bs.util.BSDataSourceCredentialsProvider_ClientProxy.getCredentials(BSDataSourceCredentialsProvider_ClientProxy.zig:157)
        at io.quarkus.agroal.runtime.AgroalVaultCredentialsProviderPassword.asProperties(AgroalVaultCredentialsProviderPassword.java:21)
        at io.agroal.api.security.AgroalDefaultSecurityProvider.getSecurityProperties(AgroalDefaultSecurityProvider.java:23)
        at io.agroal.pool.ConnectionFactory.securityProperties(ConnectionFactory.java:190)
        at io.agroal.pool.ConnectionFactory.securityProperties(ConnectionFactory.java:179)
        at io.agroal.pool.ConnectionFactory.jdbcProperties(ConnectionFactory.java:160)
        at io.agroal.pool.ConnectionFactory.createConnection(ConnectionFactory.java:204)
        at io.agroal.pool.ConnectionPool$CreateConnectionTask.call(ConnectionPool.java:470)
        at io.agroal.pool.ConnectionPool$CreateConnectionTask.call(ConnectionPool.java:452)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at io.agroal.pool.util.PriorityScheduledExecutor.beforeExecute(PriorityScheduledExecutor.java:68)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1126)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:829)

### Expected behavior

Quarkus should be able to load the JDK class javax.crypto.JceSecurity.

### Actual behavior

See stacktrace in description.

### Output of `java -version`

openjdk 11.0.11 2021-04-20 OpenJDK Runtime Environment AdoptOpenJDK-11.0.11+9 (build 11.0.11+9) OpenJDK 64-Bit Server VM AdoptOpenJDK-11.0.11+9 (build 11.0.11+9, mixed mode)

### Quarkus version or git rev

2.1.1.Final

**Comment by stuartwdouglas** [comment]
How can I reproduce this? I can load the class fine using Class.forName(). 

That error is also an initializion error, which means the static init block in JceSecurity is throwing an exception. Attaching a debugger to see what is actually throwing the exception would also be really helpful if you can't provide a reproducer.

**Comment by dhoffer** [comment]
Yeah our code project is huge, posting it is not practical.

I can say that the code that uses this is annotated with @Singleton and it
is used very early in the Quarkus framework code to decrypt the DB
password.  Our app.yml has 3 data sources but only one of them uses the
Quarkus CredentialsProvider to decrypt the password, that is the usage that
is failing.

I can try to attach a debugger to see what error is given.

I never like to use static blocks as they have too many bad behaviors.  But
since you say it's thrown there I would say there was a change in
Quarkus 2.x such that whatever is needed there is not available yet as it
works fine in Quarkus 1.x.

-Dave

**Comment by stuartwdouglas** [comment]
The static block is in the JDK itself, not much we can do about it. I tried with pretty much the exact same code and it worked.

Looking through the source code of JceSecurity I don't see how anything Quarkus was doing could actually affect it, so I am really not sure what is going on. You are going to need to attach a debugger to see what is actually throwing the exception (Java makes it really hard to debug these kinds of errors, as the cause is lost and turned into NoClassDefFoundError instead).

**Comment by dhoffer** [comment]
Understood, I will do this as soon as possible.

-Dave

**Comment by dhoffer** [comment]
I am having a hard time getting more details as I debug this.  Here is the full stack trace.

2021-08-19 15:34:14,891 ERROR [io.qua.run.Application] (main) Failed to start application (with profile prod): java.lang.NoClassDefFoundError: Could not initialize class javax.crypto.JceSecurity
        at java.base/javax.crypto.Cipher.getInstance(Cipher.java:540)
        at com.bs.util.encryption.AESEncryptDecrypt.getInitializedCipher(AESEncryptDecrypt.java:151)
        at com.bs.util.encryption.AESEncryptDecrypt.decrypt(AESEncryptDecrypt.java:127)
        at com.bs.util.encryption.AESEncryptDecryptProxy.decrypt(AESEncryptDecryptProxy.java:50)
        at com.bs.util.DataSourceCredentialsProvider.getCredentials(DataSourceCredentialsProvider.java:33)
        at com.bs.util.DataSourceCredentialsProvider_ClientProxy.getCredentials(DataSourceCredentialsProvider_ClientProxy.zig:157)
        at io.quarkus.agroal.runtime.AgroalVaultCredentialsProviderPassword.asProperties(AgroalVaultCredentialsProviderPassword.java:21)
        at io.agroal.api.security.AgroalDefaultSecurityProvider.getSecurityProperties(AgroalDefaultSecurityProvider.java:23)
        at io.agroal.pool.ConnectionFactory.securityProperties(ConnectionFactory.java:190)
        at io.agroal.pool.ConnectionFactory.securityProperties(ConnectionFactory.java:179)
        at io.agroal.pool.ConnectionFactory.jdbcProperties(ConnectionFactory.java:160)
        at io.agroal.pool.ConnectionFactory.createConnection(ConnectionFactory.java:204)
        at io.agroal.pool.ConnectionPool$CreateConnectionTask.call(ConnectionPool.java:470)
        at io.agroal.pool.ConnectionPool$CreateConnectionTask.call(ConnectionPool.java:452)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at io.agroal.pool.util.PriorityScheduledExecutor.beforeExecute(PriorityScheduledExecutor.java:68)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1126)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:829)

The line in java.base/javax.crypto.Cipher.getInstance where it is failing is on this line:
JceSecurity.canUseProvider(s.getProvider()) (line 540)

However my IDE will never go into that method it throws the NoClassDefFoundError exception.

The strange part is when I debug this the code lands on 540 about 20 times in a loop, giving an error each time.  Then it just quits with the above stack trace.

Note our app has 3 data sources and two of them use the DataSourceCredentialsProvider.  The third data source is not available in our DEV environment so that one just fails...but we expect that one to fail.  With Quarkus 2.x all of them fail with this new NoClassDefFoundError error.

I have verified this fails with the new 2.1.3.Final version as well.  

So in summary fails with all Quarkus 2.x
Works with all Quarkus 1.x
Same JDK 11
Same code base.

Here is my Java:
C:\projects\udl>java --version
openjdk 11.0.11 2021-04-20
OpenJDK Runtime Environment AdoptOpenJDK-11.0.11+9 (build 11.0.11+9)
OpenJDK 64-Bit Server VM AdoptOpenJDK-11.0.11+9 (build 11.0.11+9, mixed mode)

Note this is very early code in application startup.  Agroal is trying to get the password for our DB and to do that it has to use our AES decryption that is used in the CredentialsProvider implementation.

**Comment by stuartwdouglas** [comment]
It looks like the static init block has the following:

```
        } catch (Exception e) {
            throw new SecurityException(
                    "Can not initialize cryptographic mechanism", e);
        }
```

Can you set a breakpoint for SecurityException, and try and figure out what the underlying exception is?

Also what mode is this, is it prod, dev or both? If it is prod using fast-jar is it still a problem with legacy-jar or uber-jar?

**Comment by dhoffer** [comment]
Its not throwing in the static block, all good there.  We are running in prod using an uber-jar.

**Comment by stuartwdouglas** [comment]
Can you make a reproducer? It seems like this is all in the database startup code so I am assuming a small project with the same config would have the same issue?

**Comment by dhoffer** [comment]
I doubt that will be possible.  If no one else is having any troubles with their 2.x apps my new simple one will likely work as well.  I'll keep investigating but not getting far on this.

**Comment by stuartwdouglas** [comment]
Is there something specific in your datasource config that is triggering this code path?

It's really weird because I don't really see how Quarkus can affect a JDK class like this, especially given we don't do anything special with ClassLoading in an Uber Jar.

**Comment by dhoffer** [comment]
Nothing special.  Just using quarkus.datasource.credentials-provider config so Quarkus uses CredentialsProvider to get DB credentials.  Quarkus calls this very early as it needs the DB credentials early.

**Comment by jaikiran** [comment]
So I tried something outside of Quarkus in a standalone trivial Java code. 2 classes `Foo` and `Bar`. `Foo` just keeps calling a static method on `Bar` in a loop and catches any exception, prints it and the continues calling that method on `Bar` for the next time. This it does for 5 times. `Bar` has a static (class) initialization block and intentionally throws an exception from that block everytime. Here's what these classes look like:

Foo.java
```
public class Foo {
	public static void main(final String[] args) throws Exception {
		for (int i = 0; i < 5; i++) {
			try {
				Bar.doSomething();
			} catch(Throwable t) {
				System.err.println("Got error " + i);
				t.printStackTrace();
			}

		}

	}
}
```

Bar.java

```
public class Bar {

	static {
		System.out.println("Hi");
		if (true) {
			throw new SecurityException("intentional");
		}
	}
	
	public static void doSomething() {
		System.err.println("Done");
	}
}
```

I then run this code with Java 11 and here's what I get as the output:

```
Hi
Got error 0
java.lang.ExceptionInInitializerError
	at Foo.main(Foo.java:5)
Caused by: java.lang.SecurityException: intentional
	at Bar.<clinit>(Bar.java:6)
	... 1 more
Got error 1
java.lang.NoClassDefFoundError: Could not initialize class Bar
	at Foo.main(Foo.java:5)
Got error 2
java.lang.NoClassDefFoundError: Could not initialize class Bar
	at Foo.main(Foo.java:5)
Got error 3
java.lang.NoClassDefFoundError: Could not initialize class Bar
	at Foo.main(Foo.java:5)
Got error 4
java.lang.NoClassDefFoundError: Could not initialize class Bar
	at Foo.main(Foo.java:5)
```
Notice that the first attempt that called the static method `Bar.doSomething` which triggered the static initializer of `Bar` class fails *and* contains the full details including the underlying root cause of why it failed in its exception stacktrace:

```
Hi
Got error 0
java.lang.ExceptionInInitializerError
	at Foo.main(Foo.java:5)
Caused by: java.lang.SecurityException: intentional
	at Bar.<clinit>(Bar.java:6)
	... 1 more
```
However, the subsequent attempts on `Bar.doSomething()` just throw a `NoClassDefFoundError` without any underlying cause, like this next attempt:

```
Got error 1
java.lang.NoClassDefFoundError: Could not initialize class Bar
	at Foo.main(Foo.java:5)
```
This detail is crucial, because in one of your comment you said:

> The strange part is when I debug this the code lands on 540 about 20 times in a loop, giving an error each time. Then it just quits with the above stack trace.

So it looks like somewhere in your code in this flow you have a catch block which catches the exception (actually I think some part of your code is catching `Throwable` and not just `Exception`) but then does something else which again triggers the code which leads to this `JceSecurity` class. This it perhaps does in a loop maybe? Perhaps 20 times and then when all 20 attempts fail, it throws the last exception, which of course as seen in that above sample code will not contain the root cause.

So what you could do is, in the following flow/classes:


> at java.base/javax.crypto.Cipher.getInstance(Cipher.java:540)
at com.bs.util.encryption.AESEncryptDecrypt.getInitializedCipher(AESEncryptDecrypt.java:151)
at com.bs.util.encryption.AESEncryptDecrypt.decrypt(AESEncryptDecrypt.java:127)
at com.bs.util.encryption.AESEncryptDecryptProxy.decrypt(AESEncryptDecryptProxy.java:50)
at com.bs.util.DataSourceCredentialsProvider.getCredentials(DataSourceCredentialsProvider.java:33)

Find out where such a loop or try/catch block exists. See what's it's doing and if needs to be fixed. Since you are able to debug this code, I think if you can add a breakpoint in this catch block and inspect the first thrown exception, then it's likely that it will have a `cause` hierarchy which shows what the root cause is. If you need help understanding that root cause you can paste that exception hierarchy here (you can get it while debugger is attached by executing a `e.printStackTrace()` against that exception that you are inspecting, the stacktrace will get printed in the console of Quarkus app). It's crucial that you inspect the first exception that is thrown out of those 20 exceptions in a loop.

**Comment by stuartwdouglas** [comment]
I would also try adding a breakpoint for `ExceptionInInitializerError` and see if that helps

**Comment by dhoffer** [comment]
Thanks for the detailed reply and ideas, I will try to track this down.

I do have some initial feedback though.  We no NOT have any retries or loops in our code regarding this issue.  We make ONE call to Cipher.getInstance(cryptoModePadding) for every ONE call of Quarkus/Agroal to CredentialsProvider#getCredentials(String credentialsProviderName)

I believe the loop of 20 I mentioned is in the JDK here:

public static final Cipher getInstance(String transformation){
...
Iterator<Service> t = services.iterator();
Exception failure = null;
while (t.hasNext()) {
    Service s = t.next();
    if (JceSecurity.canUseProvider(s.getProvider()) == false) {
        continue;
    }
...

**Comment by jaikiran** [comment]
> I believe the loop of 20 I mentioned is in the JDK here:

> public static final Cipher getInstance(String transformation){
...
Iterator t = services.iterator();
Exception failure = null;
while (t.hasNext()) {
Service s = t.next();
if (JceSecurity.canUseProvider(s.getProvider()) == false) {
continue;
}

That however doesn't explain why the thrown exception is not getting propagated out of the loop and method, the very first instance. I don't see any catch block around that call to `JceSecurity.canUseProvider`, unless I'm not looking at the right JDK version/implementation?

> We are using JDK 11.0.11+9

Where did you download and install this from?

**Comment by dhoffer** [comment]
I agree, the errors just are not making sense to me.

I don't recall the site where I got the JDK but here is what is in the
release file:
IMPLEMENTOR="AdoptOpenJDK"
IMPLEMENTOR_VERSION="AdoptOpenJDK-11.0.11+9"
JAVA_VERSION="11.0.11"
JAVA_VERSION_DATE="2021-04-20"
MODULES="java.base java.compiler java.datatransfer java.xml java.prefs
java.desktop java.instrument java.logging java.management
java.security.sasl java.naming java.rmi java.management.rmi java.net.http
java.scripting java.security.jgss java.transaction.xa java.sql
java.sql.rowset java.xml.crypto java.se java.smartcardio jdk.accessibility
jdk.internal.vm.ci jdk.management jdk.unsupported jdk.internal.vm.compiler
jdk.aot jdk.internal.jvmstat jdk.attach jdk.charsets jdk.compiler
jdk.crypto.ec jdk.crypto.cryptoki jdk.crypto.mscapi jdk.dynalink
jdk.internal.ed jdk.editpad jdk.hotspot.agent jdk.httpserver
jdk.internal.le jdk.internal.opt jdk.internal.vm.compiler.management
jdk.jartool jdk.javadoc jdk.jcmd jdk.management.agent jdk.jconsole
jdk.jdeps jdk.jdwp.agent jdk.jdi jdk.jfr jdk.jlink jdk.jshell jdk.jsobject
jdk.jstatd jdk.localedata jdk.management.jfr jdk.naming.dns jdk.naming.ldap
jdk.naming.rmi jdk.net jdk.pack jdk.rmic jdk.scripting.nashorn
jdk.scripting.nashorn.shell jdk.sctp jdk.security.auth jdk.security.jgss
jdk.unsupported.desktop jdk.xml.dom jdk.zipfs"
OS_ARCH="x86_64"
OS_NAME="Windows"
SOURCE=".:git:9a5c368512c6"
BUILD_SOURCE="git:4ad15fd"
FULL_VERSION="11.0.11+9"
SEMANTIC_VERSION="11.0.11+9"
BUILD_INFO="OS: Windows Server 2012 R2 Version: 6.3"
JVM_VARIANT="Hotspot"
JVM_VERSION="11.0.11+9"
IMAGE_TYPE="JDK"

Yesterday I also installed a later JDK 11 to see if it made any difference
but I got the same error, here is its release file contents.

IMPLEMENTOR="Oracle Corporation"
IMPLEMENTOR_VERSION="18.9"
JAVA_VERSION="11"
JAVA_VERSION_DATE="2018-09-25"
MODULES="java.base java.compiler java.datatransfer java.xml java.prefs
java.desktop java.instrument java.logging java.management
java.security.sasl java.naming java.rmi java.management.rmi java.net.http
java.scripting java.security.jgss java.transaction.xa java.sql
java.sql.rowset java.xml.crypto java.se java.smartcardio jdk.accessibility
jdk.internal.vm.ci jdk.management jdk.unsupported jdk.internal.vm.compiler
jdk.aot jdk.internal.jvmstat jdk.attach jdk.charsets jdk.compiler
jdk.crypto.ec jdk.crypto.cryptoki jdk.crypto.mscapi jdk.dynalink
jdk.internal.ed jdk.editpad jdk.hotspot.agent jdk.httpserver
jdk.internal.le jdk.internal.opt jdk.internal.vm.compiler.management
jdk.jartool jdk.javadoc jdk.jcmd jdk.management.agent jdk.jconsole
jdk.jdeps jdk.jdwp.agent jdk.jdi jdk.jfr jdk.jlink jdk.jshell jdk.jsobject
jdk.jstatd jdk.localedata jdk.management.jfr jdk.naming.dns jdk.naming.rmi
jdk.net jdk.pack jdk.rmic jdk.scripting.nashorn jdk.scripting.nashorn.shell
jdk.sctp jdk.security.auth jdk.security.jgss jdk.unsupported.desktop
jdk.xml.dom jdk.zipfs"
OS_ARCH="x86_64"
OS_NAME="Windows"
SOURCE=".:76072a077ee1"

**Comment by jaikiran** [comment]
If the "easier" path of debugging doesn't solve this, then I think the last resort is the Byteman approach. In that case, please follow these steps:

- Download Byteman from https://repo1.maven.org/maven2/org/jboss/byteman/byteman/4.0.16/byteman-4.0.16.jar and save it to some local directory on that setup
- Copy the following byteman rules into a file named `byteman-rules.btm` and save it in the same dir as the byteman jar:
```
RULE ExceptionInInitializerError no-arg
CLASS java.lang.ExceptionInInitializerError
METHOD <init>()
AT EXIT
IF TRUE
DO traceStack("*** ExceptionInInitializerError no-arg called")
ENDRULE

RULE ExceptionInInitializerError throwable-arg
CLASS java.lang.ExceptionInInitializerError
METHOD <init>(Throwable)
AT EXIT
IF TRUE
DO traceStack("*** ExceptionInInitializerError called with a cause");
   $1.printStackTrace();
ENDRULE


RULE ExceptionInInitializerError string-arg
CLASS java.lang.ExceptionInInitializerError
METHOD <init>(String)
AT EXIT
IF TRUE
DO traceStack("*** ExceptionInInitializerError with message called")
ENDRULE


RULE JceSecurity exception exit
CLASS javax.crypto.JceSecurity
METHOD <clinit>
AT EXCEPTION EXIT
IF TRUE
DO traceStack("*** JceSecurity failed with exception");
  $^.printStackTrace();
ENDRULE




```

- Assuming that location you saved these files is "C:\projects\udl\byteman\", the launch your Quarkus application using:

```
java -javaagent:"C:\projects\udl\byteman\byteman-4.0.16.jar=script:C:\projects\udl\byteman\byteman-rules.btm,boot:C:\projects\udl\byteman\byteman-4.0.16.jar,prop:org.jboss.byteman.verbose=true,prop:org.jboss.byteman.transform.all=true" -jar <the-quarkus-application-jar>
```

(you might have to fiddle with the java launch command. Essentially you just need to additionally add the javaagent option as noted above in addition to whatever you are currently using to launch your application)

This should generate (hopefully) enough exception stacktrace/output on the console when the JceSecurity class initialization errors out.

P.S: Long back, I had wondered whether Quarkus should have a much more seamless integration with Byteman to help with issues like these. I think it's time to experiment a bit and see how to make debugging Quarkus in cases like these easier with Byteman. I'll see if I can try a PoC to make this easier.

**Comment by dhoffer** [comment]
Fyi...I tried 2.1.3.Final in 3 more Quarkus apps, all much smaller then the first one.  Two did not have this error but one does so 2 in total have this error and 2 do not.  And just as a reminder this is with 2.x only, 1.x is fine.

**Comment by geoand** [comment]
Is this still an issue?

**Comment by geoand** [comment]
Closing for lack of feedback
